File: go/auth_test.go
package keycloak

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strconv"
	"strings"
	"testing"
	"time"

	jose "github.com/go-jose/go-jose/v4"
	"github.com/go-jose/go-jose/v4/jwt"
)

// authFixture spins up an httptest server standing in for a realm's OIDC
// endpoints and returns an AuthClient pointed at it.
type authFixture struct {
	srv    *httptest.Server
	auth   *AuthClient
	lastFn func(path string, form url.Values)
}

func newAuthFixture(t *testing.T, secret string) *authFixture {
	t.Helper()
	f := &authFixture{}
	mux := http.NewServeMux()
	base := "/realms/test/protocol/openid-connect"
	mux.HandleFunc(base+"/token", func(w http.ResponseWriter, r *http.Request) {
		_ = r.ParseForm()
		if f.lastFn != nil {
			f.lastFn("token", r.Form)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"access_token":"AT","token_type":"Bearer","expires_in":300,` +
			`"refresh_token":"RT","scope":"openid","id_token":"IDT"}`))
	})
	mux.HandleFunc(base+"/token/introspect", func(w http.ResponseWriter, r *http.Request) {
		_ = r.ParseForm()
		if f.lastFn != nil {
			f.lastFn("introspect", r.Form)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"active":true,"username":"svc","client_id":"it-client","sub":"u1"}`))
	})
	mux.HandleFunc(base+"/logout", func(w http.ResponseWriter, r *http.Request) {
		_ = r.ParseForm()
		if f.lastFn != nil {
			f.lastFn("logout", r.Form)
		}
		w.WriteHeader(http.StatusNoContent)
	})
	f.srv = httptest.NewServer(mux)
	t.Cleanup(f.srv.Close)
	cfg := Config{ServerURL: f.srv.URL, Realm: "test", ClientID: "app", ClientSecret: secret,
		ReadTimeout: 30000}.withDefaults()
	f.auth = newAuthClient(cfg, nil)
	return f
}

func TestCreateAuthorizationRequest(t *testing.T) {
	cfg := Config{ServerURL: "https://kc.example.com", Realm: "demo", ClientID: "app"}.withDefaults()
	a := newAuthClient(cfg, nil)
	req := a.CreateAuthorizationRequest("https://app/cb")
	u, err := url.Parse(req.URL)
	if err != nil {
		t.Fatalf("bad URL: %v", err)
	}
	q := u.Query()
	if u.Scheme+"://"+u.Host+u.Path != "https://kc.example.com/realms/demo/protocol/openid-connect/auth" {
		t.Fatalf("endpoint: %s", u.String())
	}
	for k, want := range map[string]string{
		"response_type": "code", "client_id": "app", "redirect_uri": "https://app/cb",
		"scope": "openid", "code_challenge_method": "S256", "state": req.State, "nonce": req.Nonce,
	} {
		if q.Get(k) != want {
			t.Errorf("%s = %q, want %q", k, q.Get(k), want)
		}
	}
	sum := sha256.Sum256([]byte(req.CodeVerifier))
	if q.Get("code_challenge") != base64.RawURLEncoding.EncodeToString(sum[:]) {
		t.Errorf("code_challenge must be base64url(sha256(verifier))")
	}
	// Distinct per call.
	b := a.CreateAuthorizationRequest("https://app/cb")
	if req.CodeVerifier == b.CodeVerifier || req.State == b.State || req.Nonce == b.Nonce {
		t.Error("verifier/state/nonce must differ per call")
	}
}

func TestClientCredentialsToken(t *testing.T) {
	f := newAuthFixture(t, "sekret")
	ts, err := f.auth.ClientCredentialsToken(context.Background())
	if err != nil {
		t.Fatalf("client credentials: %v", err)
	}
	if ts.AccessToken != "AT" || ts.RefreshToken != "RT" || ts.IDToken != "IDT" {
		t.Fatalf("token mapping: access=%q refresh=%q id=%q", ts.AccessToken, ts.RefreshToken, ts.IDToken)
	}
	if ts.ExpiresIn <= 0 || ts.ExpiresAt <= 0 {
		t.Fatalf("expiry mapping: ExpiresIn=%d ExpiresAt=%d", ts.ExpiresIn, ts.ExpiresAt)
	}
}

func TestExchangeCodeAndRefresh(t *testing.T) {
	f := newAuthFixture(t, "sekret")
	var form url.Values
	f.lastFn = func(path string, fm url.Values) {
		if path == "token" {
			form = fm
		}
	}

	ts, err := f.auth.ExchangeCode(context.Background(), "the-code", "https://app/cb", "verifier", "")
	if err != nil || ts.AccessToken != "AT" {
		t.Fatalf("exchange: %+v %v", ts, err)
	}
	if form.Get("grant_type") != "authorization_code" || form.Get("code") != "the-code" ||
		form.Get("code_verifier") != "verifier" || form.Get("redirect_uri") != "https://app/cb" {
		t.Fatalf("exchange must send authorization_code + code + PKCE verifier + redirect_uri: %v", form)
	}

	ts2, err := f.auth.Refresh(context.Background(), "old-rt")
	if err != nil || ts2.AccessToken != "AT" {
		t.Fatalf("refresh: %+v %v", ts2, err)
	}
	if form.Get("grant_type") != "refresh_token" || form.Get("refresh_token") != "old-rt" {
		t.Fatalf("refresh must send refresh_token grant: %v", form)
	}
}

// signIDToken produces an RS256-signed id_token carrying a nonce claim, for the
// ExchangeCode nonce-replay tests.
func signIDToken(t *testing.T, key *rsa.PrivateKey, kid, iss, aud, nonce string) string {
	t.Helper()
	sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: key},
		(&jose.SignerOptions{}).WithType("JWT").WithHeader("kid", kid))
	if err != nil {
		t.Fatalf("signer: %v", err)
	}
	std := jwt.Claims{Subject: "user1", Issuer: iss, Audience: jwt.Audience{aud},
		Expiry: jwt.NewNumericDate(time.Now().Add(5 * time.Minute)), IssuedAt: jwt.NewNumericDate(time.Now())}
	s, err := jwt.Signed(sig).Claims(std).Claims(map[string]any{"nonce": nonce}).Serialize()
	if err != nil {
		t.Fatalf("serialize: %v", err)
	}
	return s
}

// TestExchangeCodeNonceValidation proves that when an expectedNonce is supplied,
// ExchangeCode fully signature-validates the returned id_token and rejects a
// mismatched / missing nonce (OIDC nonce replay protection). Empty nonce skips it.
func TestExchangeCodeNonceValidation(t *testing.T) {
	priv, _ := rsa.GenerateKey(rand.Reader, 2048)
	jwks := jose.JSONWebKeySet{Keys: []jose.JSONWebKey{
		{Key: &priv.PublicKey, KeyID: "k1", Algorithm: "RS256", Use: "sig"},
	}}
	base := "/realms/test/protocol/openid-connect"
	var issuer string
	includeIDToken := true
	mux := http.NewServeMux()
	mux.HandleFunc(base+"/token", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		if includeIDToken {
			idt := signIDToken(t, priv, "k1", issuer, "app", "server-nonce")
			_, _ = w.Write([]byte(`{"access_token":"AT","token_type":"Bearer","expires_in":300,"id_token":"` + idt + `"}`))
		} else {
			_, _ = w.Write([]byte(`{"access_token":"AT","token_type":"Bearer","expires_in":300}`))
		}
	})
	mux.HandleFunc(base+"/certs", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(jwks)
	})
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)
	issuer = srv.URL + "/realms/test"

	v := newValidator(validatorOptions{
		jwksURI: srv.URL + base + "/certs", issuer: issuer, audience: "app",
		allowedAlgs: []jose.SignatureAlgorithm{jose.RS256}, clockSkewSec: 30, minRefetch: time.Minute,
	})
	cfg := Config{ServerURL: srv.URL, Realm: "test", ClientID: "app", ClientSecret: "sekret",
		ReadTimeout: 30000}.withDefaults()
	auth := newAuthClient(cfg, v)
	ctx := context.Background()

	// Matching nonce → accepted.
	ts, err := auth.ExchangeCode(ctx, "code", srv.URL+"/cb", "verifier", "server-nonce")
	if err != nil || ts.AccessToken != "AT" {
		t.Fatalf("matching nonce must pass: %+v %v", ts, err)
	}

	// Mismatched nonce → rejected as *AuthError.
	_, err = auth.ExchangeCode(ctx, "code", srv.URL+"/cb", "verifier", "attacker-nonce")
	var ae *AuthError
	if !errors.As(err, &ae) {
		t.Fatalf("mismatched nonce must yield *AuthError, got %v", err)
	}

	// Missing id_token while a nonce is expected → rejected (fail-closed).
	includeIDToken = false
	_, err = auth.ExchangeCode(ctx, "code", srv.URL+"/cb", "verifier", "server-nonce")
	if !errors.As(err, &ae) {
		t.Fatalf("missing id_token with expected nonce must yield *AuthError, got %v", err)
	}
}

func TestIntrospect(t *testing.T) {
	f := newAuthFixture(t, "sekret")
	r, err := f.auth.Introspect(context.Background(), "tok")
	if err != nil {
		t.Fatalf("introspect: %v", err)
	}
	if !r.Active || r.Username != "svc" || r.ClientID != "it-client" || r.Claims["sub"] != "u1" {
		t.Fatalf("introspection result: %+v", r)
	}
}

func TestLogoutPostsCredentials(t *testing.T) {
	f := newAuthFixture(t, "sekret")
	var got url.Values
	f.lastFn = func(path string, form url.Values) {
		if path == "logout" {
			got = form
		}
	}
	if err := f.auth.Logout(context.Background(), "RT"); err != nil {
		t.Fatalf("logout: %v", err)
	}
	if got.Get("refresh_token") != "RT" || got.Get("client_id") != "app" || got.Get("client_secret") != "sekret" {
		t.Fatalf("logout form: %v", got)
	}
}

func TestClientCredentialsErrorMapping(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		_, _ = w.Write([]byte(`{"error":"invalid_client","error_description":"bad"}`))
	}))
	t.Cleanup(srv.Close)
	// The server has no realm path; point token URL straight at it via a realm that maps to "/".
	cfg := Config{ServerURL: srv.URL, Realm: "test", ClientID: "app", ClientSecret: "x"}.withDefaults()
	a := newAuthClient(cfg, nil)
	_, err := a.ClientCredentialsToken(context.Background())
	var ae *AuthError
	if !errors.As(err, &ae) {
		t.Fatalf("expected *AuthError, got %v", err)
	}
	if ae.OAuthError != "invalid_client" {
		t.Fatalf("OAuthError = %q, want invalid_client", ae.OAuthError)
	}
}

func TestLogoutErrorStatus(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
	}))
	t.Cleanup(srv.Close)
	cfg := Config{ServerURL: srv.URL, Realm: "test", ClientID: "app", ClientSecret: "x"}.withDefaults()
	a := newAuthClient(cfg, nil)
	if err := a.Logout(context.Background(), "RT"); err == nil || !strings.Contains(err.Error(), "HTTP 400") {
		t.Fatalf("logout must surface HTTP 400: %v", err)
	}
}

// postForm 은 `>= 400` 만 실패로 봤다. SDK 는 리다이렉트를 일부러 따라가지 않고
// (`noFollowRedirect`) 3xx 를 **호출자에게 그대로 올리므로**, 그 판정은 302 를 성공으로 읽는다.
// 결과: 세션이 살아 있는데 Logout 이 nil 을 돌려주고, Introspect 는 빈 본문을 파싱한다.
// 성공은 2xx 다 — 그 밖은 전부 오류로 닫는다.
func TestPostFormRejectsNon2xx(t *testing.T) {
	// ⚠️ 1xx 는 넣지 않는다 — net/http 를 통해서는 도달할 수 없다. 실측(probe): 핸들러가
	// `WriteHeader(100)` 을 써도 전송 계층이 informational 응답을 소비하고 클라이언트가 보는
	// 최종 상태는 **200** 이다. 넣으면 코드가 아니라 테스트가 틀린 채로 빨개진다.
	for _, code := range []int{
		http.StatusMovedPermanently, http.StatusFound, http.StatusSeeOther,
		http.StatusTemporaryRedirect, http.StatusPermanentRedirect,
	} {
		t.Run(http.StatusText(code), func(t *testing.T) {
			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
				w.Header().Set("Location", "/elsewhere")
				w.WriteHeader(code)
			}))
			t.Cleanup(srv.Close)
			cfg := Config{ServerURL: srv.URL, Realm: "test", ClientID: "app", ClientSecret: "x"}.withDefaults()
			a := newAuthClient(cfg, nil)

			err := a.Logout(context.Background(), "RT")
			if err == nil {
				t.Fatalf("Logout 이 HTTP %d 를 성공으로 읽었다 — 세션이 살아 있는데 nil 을 돌려준다", code)
			}
			if !strings.Contains(err.Error(), "HTTP "+strconv.Itoa(code)) {
				t.Fatalf("오류는 실제 상태코드를 담아야 한다(HTTP %d): %v", code, err)
			}
		})
	}
}

File: node/test/unit/auth.test.ts
import { createHash } from 'node:crypto'
import { describe, it, expect, vi, beforeAll, beforeEach } from 'vitest'

// openid-client는 네트워크 경계 — 함수형 API를 목킹해 매핑/PKCE/예외변환을 격리 검증한다.
vi.mock('openid-client', () => ({
  discovery: vi.fn(),
  clientCredentialsGrant: vi.fn(),
  authorizationCodeGrant: vi.fn(),
  refreshTokenGrant: vi.fn(),
  tokenIntrospection: vi.fn(),
  allowInsecureRequests: vi.fn(),
  customFetch: Symbol('customFetch'),
}))

import { generateKeyPair, exportJWK, SignJWT, createLocalJWKSet, type JWTVerifyGetKey } from 'jose'
import * as oidc from 'openid-client'
import { AuthClient } from '../../src/auth.js'
import { defineConfig, type KeycloakConfigInput } from '../../src/config.js'
import {
  KeycloakAuthError,
  KeycloakTokenValidationError,
  KeycloakTransportError,
} from '../../src/errors.js'
import { JwtValidator } from '../../src/jwt.js'
import { TokenSet } from '../../src/tokens.js'

const cfg = defineConfig({
  serverUrl: 'https://kc.example.com',
  realm: 'demo',
  clientId: 'app',
  clientSecret: 'sekret',
})

const AUTH_EP = 'https://kc.example.com/realms/demo/protocol/openid-connect/auth'
const LOGOUT_EP = 'https://kc.example.com/realms/demo/protocol/openid-connect/logout'

beforeEach(() => {
  vi.clearAllMocks()
  // 대부분의 토큰 연산은 discovery 결과(Configuration)를 필요로 한다. timeout은 setter 가능.
  vi.mocked(oidc.discovery).mockResolvedValue({ timeout: 0 } as never)
})

describe('discovery 실패 경계 변환 (§4 — 원시 하위 오류 누출 금지)', () => {
  it('전송 실패(undici fetch failed)는 KeycloakTransportError로 변환한다', async () => {
    vi.mocked(oidc.discovery).mockRejectedValue(
      new TypeError('fetch failed', {
        cause: Object.assign(new Error('connect ECONNREFUSED'), { code: 'ECONNREFUSED' }),
      }),
    )
    await expect(new AuthClient(cfg).clientCredentialsToken()).rejects.toBeInstanceOf(
      KeycloakTransportError,
    )
  })

  it('타임아웃(AbortError)은 KeycloakTransportError로 변환한다', async () => {
    vi.mocked(oidc.discovery).mockRejectedValue(
      Object.assign(new Error('aborted'), { name: 'AbortError' }),
    )
    await expect(new AuthClient(cfg).refresh('rt')).rejects.toBeInstanceOf(KeycloakTransportError)
  })

  it('비전송 실패(불량 메타데이터 등)는 KeycloakAuthError로 변환한다', async () => {
    vi.mocked(oidc.discovery).mockRejectedValue(new Error('bad OIDC metadata'))
    await expect(new AuthClient(cfg).introspect('tok')).rejects.toBeInstanceOf(KeycloakAuthError)
  })

  it('readTimeoutMs(초)를 discovery 옵션 timeout으로 전달한다(최초 well-known 페치에 적용)', async () => {
    const c = defineConfig({
      serverUrl: 'https://kc',
      realm: 'r',
      clientId: 'c',
      clientSecret: 's',
      readTimeoutMs: 5000,
    })
    vi.mocked(oidc.clientCredentialsGrant).mockResolvedValue({
      access_token: 'x',
      expires_in: 1,
    } as never)
    await new AuthClient(c).clientCredentialsToken()
    expect(vi.mocked(oidc.discovery).mock.calls.at(-1)?.[4]).toMatchObject({ timeout: 5 })
  })
})

describe('createAuthorizationRequest (동기·네트워크 없음)', () => {
  it('PKCE S256 인가 URL을 모든 필수 파라미터와 함께 조립한다', () => {
    const req = new AuthClient(cfg).createAuthorizationRequest('https://app.example.com/cb')
    const u = new URL(req.url)
    expect(u.origin + u.pathname).toBe(AUTH_EP)
    expect(u.searchParams.get('response_type')).toBe('code')
    expect(u.searchParams.get('client_id')).toBe('app')
    expect(u.searchParams.get('redirect_uri')).toBe('https://app.example.com/cb')
    expect(u.searchParams.get('scope')).toBe('openid')
    expect(u.searchParams.get('code_challenge_method')).toBe('S256')
    expect(u.searchParams.get('state')).toBe(req.state)
    expect(u.searchParams.get('nonce')).toBe(req.nonce)
    // code_challenge === base64url(sha256(code_verifier)) — RFC 7636 S256
    const expected = createHash('sha256').update(req.codeVerifier).digest().toString('base64url')
    expect(u.searchParams.get('code_challenge')).toBe(expected)
  })

  it('설정된 scopes가 있으면 그대로 사용한다', () => {
    const c = defineConfig({
      serverUrl: 'https://kc',
      realm: 'r',
      clientId: 'c',
      scopes: ['openid', 'profile'],
    })
    const req = new AuthClient(c).createAuthorizationRequest('https://x/cb')
    expect(new URL(req.url).searchParams.get('scope')).toBe('openid profile')
  })

  it('호출마다 verifier/state/nonce가 서로 다르다(재사용 금지)', () => {
    const auth = new AuthClient(cfg)
    const a = auth.createAuthorizationRequest('https://x/cb')
    const b = auth.createAuthorizationRequest('https://x/cb')
    expect(a.codeVerifier).not.toBe(b.codeVerifier)
    expect(a.state).not.toBe(b.state)
    expect(a.nonce).not.toBe(b.nonce)
  })
})

describe('clientCredentialsToken', () => {
  it('토큰 응답(snake_case)을 TokenSet으로 매핑한다', async () => {
    vi.mocked(oidc.clientCredentialsGrant).mockResolvedValue({
      access_token: 'AT',
      token_type: 'Bearer',
      expires_in: 300,
      refresh_token: 'RT',
      scope: 'openid',
    } as never)
    const ts = await new AuthClient(cfg).clientCredentialsToken()
    expect(ts).toBeInstanceOf(TokenSet)
    expect(ts.accessToken).toBe('AT')
    expect(ts.expiresIn).toBe(300)
    expect(ts.refreshToken).toBe('RT')
  })

  it('설정된 scopes를 grant에 전달한다', async () => {
    const c = defineConfig({
      serverUrl: 'https://kc',
      realm: 'r',
      clientId: 'c',
      clientSecret: 's',
      scopes: ['a', 'b'],
    })
    vi.mocked(oidc.clientCredentialsGrant).mockResolvedValue({
      access_token: 'x',
      expires_in: 1,
    } as never)
    await new AuthClient(c).clientCredentialsToken()
    expect(vi.mocked(oidc.clientCredentialsGrant).mock.calls.at(-1)?.[1]).toEqual({ scope: 'a b' })
  })

  it('grant 실패를 KeycloakAuthError로 변환한다(하위 타입 은닉)', async () => {
    vi.mocked(oidc.clientCredentialsGrant).mockRejectedValue(new Error('invalid_client'))
    await expect(new AuthClient(cfg).clientCredentialsToken()).rejects.toBeInstanceOf(
      KeycloakAuthError,
    )
  })
})

describe('exchangeCode', () => {
  it('code+verifier로 토큰을 교환하고, 콜백 URL에 code+iss를 붙여 넘긴다', async () => {
    vi.mocked(oidc.authorizationCodeGrant).mockResolvedValue({
      access_token: 'AT2',
      expires_in: 60,
    } as never)
    const ts = await new AuthClient(cfg).exchangeCode('the-code', 'https://app/cb', 'verifier')
    expect(ts.accessToken).toBe('AT2')
    const [, currentUrl, checks] = vi.mocked(oidc.authorizationCodeGrant).mock.calls.at(-1)!
    expect((currentUrl as URL).searchParams.get('code')).toBe('the-code')
    expect((currentUrl as URL).searchParams.get('iss')).toBe('https://kc.example.com/realms/demo')
    expect(checks).toEqual({ pkceCodeVerifier: 'verifier' })
  })

  it('nonce를 넘기면 expectedNonce로 전달한다(openid-client가 id_token nonce를 검증하도록)', async () => {
    vi.mocked(oidc.authorizationCodeGrant).mockResolvedValue({
      access_token: 'AT2',
      expires_in: 60,
    } as never)
    await new AuthClient(cfg).exchangeCode('c', 'https://app/cb', 'verifier', 'the-nonce')
    const [, , checks] = vi.mocked(oidc.authorizationCodeGrant).mock.calls.at(-1)!
    expect(checks).toEqual({ pkceCodeVerifier: 'verifier', expectedNonce: 'the-nonce' })
  })

  it('실패를 KeycloakAuthError로 변환한다', async () => {
    vi.mocked(oidc.authorizationCodeGrant).mockRejectedValue(new Error('invalid_grant'))
    await expect(
      new AuthClient(cfg).exchangeCode('c', 'https://app/cb', 'v'),
    ).rejects.toBeInstanceOf(KeycloakAuthError)
  })
})

describe('refresh', () => {
  it('refresh_token grant 결과를 TokenSet으로 매핑한다', async () => {
    vi.mocked(oidc.refreshTokenGrant).mockResolvedValue({
      access_token: 'AT3',
      expires_in: 120,
      refresh_token: 'RT3',
    } as never)
    const ts = await new AuthClient(cfg).refresh('old-rt')
    expect(ts.accessToken).toBe('AT3')
    expect(ts.refreshToken).toBe('RT3')
    expect(vi.mocked(oidc.refreshTokenGrant).mock.calls.at(-1)?.[1]).toBe('old-rt')
  })
})

describe('introspect', () => {
  it('introspection 응답을 {active, username, clientId, claims}로 매핑한다', async () => {
    vi.mocked(oidc.tokenIntrospection).mockResolvedValue({
      active: true,
      sub: 'u1',
      username: 'bob',
      client_id: 'it-client',
    } as never)
    const r = await new AuthClient(cfg).introspect('tok')
    expect(r.active).toBe(true)
    expect(r.username).toBe('bob')
    expect(r.clientId).toBe('it-client')
    expect(r.claims['username']).toBe('bob')
  })

  it('active=false를 그대로 반영한다', async () => {
    vi.mocked(oidc.tokenIntrospection).mockResolvedValue({ active: false } as never)
    const r = await new AuthClient(cfg).introspect('tok')
    expect(r.active).toBe(false)
  })
})

describe('logout (백채널 — refresh_token revoke)', () => {
  it('end_session 엔드포인트에 refresh_token+client 인증을 POST한다', async () => {
    const fetchSpy = vi
      .spyOn(globalThis, 'fetch')
      .mockResolvedValue(new Response(null, { status: 204 }))
    await new AuthClient(cfg).logout('RT')
    const [url, init] = fetchSpy.mock.calls[0]!
    expect(String(url)).toBe(LOGOUT_EP)
    expect(init?.method).toBe('POST')
    const body = init!.body as URLSearchParams
    expect(body.get('refresh_token')).toBe('RT')
    expect(body.get('client_id')).toBe('app')
    expect(body.get('client_secret')).toBe('sekret')
    fetchSpy.mockRestore()
  })

  it('비정상 응답(HTTP 4xx/5xx)을 KeycloakAuthError로 변환한다', async () => {
    const fetchSpy = vi
      .spyOn(globalThis, 'fetch')
      .mockResolvedValue(new Response(null, { status: 400 }))
    await expect(new AuthClient(cfg).logout('RT')).rejects.toBeInstanceOf(KeycloakAuthError)
    fetchSpy.mockRestore()
  })
})

describe('validate — expectedAudience', () => {
  let priv: Awaited<ReturnType<typeof generateKeyPair>>['privateKey']
  let keys: JWTVerifyGetKey

  beforeAll(async () => {
    const kp = await generateKeyPair('RS256')
    priv = kp.privateKey
    const jwk = await exportJWK(kp.publicKey)
    keys = createLocalJWKSet({ keys: [{ ...jwk, kid: 'k1', use: 'sig', alg: 'RS256' }] })
  })

  const sign = (aud: string) =>
    new SignJWT({ sub: 'u', aud })
      .setProtectedHeader({ alg: 'RS256', kid: 'k1' })
      .setIssuedAt()
      .setIssuer('https://kc.example.com/realms/demo')
      .setExpirationTime('5m')
      .sign(priv)

  /** 원격 JWKS 대신 로컬 키셋을 쓰는 검증기로 조립한다 — 실제 aud 판정을 네트워크 없이 관찰. */
  const authWithLocalJwks = (input: KeycloakConfigInput): AuthClient => {
    const spy = vi
      .spyOn(JwtValidator, 'forJwksUri')
      .mockImplementation((_uri, opts) => new JwtValidator(keys, opts))
    try {
      return new AuthClient(defineConfig(input))
    } finally {
      spy.mockRestore()
    }
  }

  const base: KeycloakConfigInput = {
    serverUrl: 'https://kc.example.com',
    realm: 'demo',
    clientId: 'app',
  }

  it('미지정이면 clientId를 기대한다(기존 동작 유지)', async () => {
    const auth = authWithLocalJwks(base)
    await expect(auth.validate(await sign('app'))).resolves.toMatchObject({ audience: ['app'] })
    await expect(auth.validate(await sign('some-api'))).rejects.toBeInstanceOf(
      KeycloakTokenValidationError,
    )
  })

  it('설정하면 그 값을 기대하고 clientId만 담긴 토큰은 거부한다', async () => {
    // 기본 realm은 client-credentials 토큰 aud에 client_id를 넣지 않는다(audience 매퍼 필요) —
    // 리소스 서버처럼 API 이름이 aud인 경우를 위한 탈출구.
    const auth = authWithLocalJwks({ ...base, expectedAudience: 'some-api' })
    await expect(auth.validate(await sign('some-api'))).resolves.toMatchObject({
      audience: ['some-api'],
    })
    await expect(auth.validate(await sign('app'))).rejects.toBeInstanceOf(
      KeycloakTokenValidationError,
    )
  })
})

describe('validate', () => {
  it('주입된 JwtValidator에 위임한다(auth는 검증 로직을 재구현하지 않는다)', async () => {
    const validated = { subject: 'u', audience: ['app'], issuer: 'iss', claims: {} }
    const fakeValidator = { validate: vi.fn().mockResolvedValue(validated) }
    const auth = new AuthClient(cfg, fakeValidator as unknown as JwtValidator)
    await expect(auth.validate('tok')).resolves.toBe(validated)
    expect(fakeValidator.validate).toHaveBeenCalledWith('tok')
  })
})
